extensions/opencv/MotionDetector.cpp

```diff
@@ -0,0 +1,222 @@
+}
+
+void MotionDetector::onSchedule(core::ProcessContext *context, core::ProcessSessionFactory *sessionFactory) {
+  std::string value;
+
+  if (context->getProperty(ImageEncoding.getName(), value)) {
+    image_encoding_ = value;
+  }
+
+  if (context->getProperty(MinInterestArea.getName(), value)) {
+    core::Property::StringToInt(value, min_area_);
+  }
+
+  if (context->getProperty(Threshold.getName(), value)) {
+    core::Property::StringToInt(value, threshold_);
+  }
+
+  if (context->getProperty(DilateIter.getName(), value)) {
+    core::Property::StringToInt(value, dil_iter_);
+  }
+
+  if (context->getProperty(BackgroundFrame.getName(), value) && !value.empty()) {
+    bg_img_ = cv::imread(value, cv::IMREAD_GRAYSCALE);
+    double scale = 500.0 / bg_img_.size().width;
+    cv::resize(bg_img_, bg_img_, cv::Size(0, 0), scale, scale);
+    cv::GaussianBlur(bg_img_, bg_img_, cv::Size(21, 21), 0, 0);
+    bg_img_.convertTo(background_, CV_32F);
+  }
+
+  if (context->getProperty(PrevAsBackground.getName(), value)) {
+    update_bg_ = (value == "yes");
+  }
+
+  logger_->log_trace("MotionDetector processor scheduled");
+}
+
+bool MotionDetector::detectAndDraw(cv::Mat &frame) {
+  cv::Mat gray;
+  cv::Mat img_diff, thresh;
+  std::vector<cv::Mat> contours;
+
+  logger_->log_debug("Detect and Draw");
+
+  cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
+  cv::GaussianBlur(gray, gray, cv::Size(21, 21), 0, 0);
+
+  // Get difference between current frame and background
+  logger_->log_debug("Get difference [%d x %d] [%d x %d]", bg_img_.rows, bg_img_.cols, gray.rows, gray.cols);
+  cv::absdiff(gray, bg_img_, img_diff);
+  logger_->log_debug("Apply threshold");
+  cv::threshold(img_diff, thresh, threshold_, 255, cv::THRESH_BINARY);
+  // Image processing.
+  logger_->log_debug("Dilation");
+  cv::dilate(thresh, thresh, cv::Mat(), cv::Point(-1, -1), dil_iter_);
+  cv::findContours(thresh, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
+
+  // Finish process
+  logger_->log_debug("Draw contours");
+  bool moved = false;
+  for (const auto &contour : contours) {
+    auto area = cv::contourArea(contour);
+    if (area < min_area_) {
+      continue;
+    }
+    // TODO - Nghia: If a contour exists for >n frames, then we should consider it as background
+    // TODO - Nghia: How? Map<rectangle, cnt>?
+    moved = true;
+    cv::Rect bbox = cv::boundingRect(contour);
+    cv::rectangle(frame, bbox.tl(), bbox.br(), cv::Scalar(0, 255, 0), 1, 8, 0);
+  }
+  logger_->log_debug("Updating bg");
+  if (update_bg_) {
+    // Use the previous frame as background.
+    logger_->log_debug("Prev as bg");
+    gray.copyTo(bg_img_);
+  } else if (!moved) {
+    logger_->log_debug("Not moved");
+    // Adaptive background, update background so that the illumnation does not affect that much.
+    cv::accumulateWeighted(gray, background_, 0.5);
+    cv::convertScaleAbs(background_, bg_img_);
+  }
+  logger_->log_debug("Finish Detect and Draw");
```

Comment:
log_trace please for all the log messages that doesn't indicate any decision, just the beginning/end of an action.